crates/walrus-service/src/node/blob_event_processor.rs

```diff
@@ -132,39 +138,68 @@ impl BackgroundEventProcessor {
-            walrus_utils::with_label!(
-                self.node.metrics.pending_processing_blob_event_in_queue,
-                &self.worker_index.to_string()
-            )
-            .dec();
-
-            let TrackedEvent {
-                event_handle,
-                blob_event,
-                checkpoint_position,
-                _guard,
-            } = tracked_event;
-
-            // The guard will automatically decrement the counter when dropped.
-            if let Err(error) = self
-                .process_event(event_handle, blob_event, checkpoint_position)
-                .await
-            {
-                // TODO(WAL-874): to keep the same behavior as before BackgroundEventProcessor, we
-                // should propagate the error to the node and exit the process if necessary.
-                tracing::error!(?error, "error processing blob event");
+        loop {
+            tokio::select! {
+                _ = self.cancel_token.cancelled() => {
+                    tracing::info!(
+                        worker_index = self.worker_index,
+                        "background event processor shutting down gracefully"
+                    );
+                    break;
+                }
+                Some(tracked_event) = self.event_receiver.recv() => {
+                    walrus_utils::with_label!(
+                        self.node.metrics.pending_processing_blob_event_in_queue,
+                        &self.worker_index.to_string()
+                    )
+                    .dec();
+
+                    let TrackedEvent {
+                        event_handle,
+                        blob_event,
+                        checkpoint_position,
+                        _guard,
+                    } = tracked_event;
+
+                    // The guard will automatically decrement the counter when dropped.
+                    if let Err(error) = self
+                        .process_event(event_handle, blob_event, checkpoint_position)
+                        .await
+                    {
+                        // Propagate error to the node by triggering node-wide shutdown. All event
+                        // processing errors should cause the node to exit.
+                        tracing::error!(
+                            ?error,
+                            worker_index = self.worker_index,
+                            "error processing blob event, triggering node shutdown"
+                        );
+                        self.node_cancel_token.cancel();
+                        break;
+                    }
+                }
+                else => {
+                    // Channel closed, exit gracefully
+                    tracing::info!(
+                        worker_index = self.worker_index,
+                        "background event processor channel closed, shutting down"
+                    );
+                    break;
```

Comment:
I wonder if in None branch (i.e. event channel closed), we should also trip the `node_cancel_token` to trigger shutdown.